### DISCUSSION TITLE
Reduce jar size

### DISCUSSION THREAD

**Discussion by user50** [deepjavalibrary/djl#2566] [Q&A] [answered]
Can someone suggest ways to reduce the size of my fat JAR file that is currently around 1.2 GB? I am using the Shadow JAR plugin in my Gradle file to create the fat JAR. I believe this large file size is due to the inclusion of the PyTorch library, which is quite large, in my dependencies:

```
runtimeOnly "ai.djl.pytorch:pytorch-native-cu117:1.13.0:linux-x86_64"
runtimeOnly "ai.djl.pytorch:pytorch-jni:1.13.0-0.20.0"
```

Are there any alternative approaches that I could use to achieve the same functionality but with a smaller JAR file size?

**Comment by frankfliu** [comment, marked as answer]
The pytorch GPU binary itself is more than 1G. I don't think there is way to reduce it's size if you must use GPU and want to distribute it.

Here is a few options:
1. use DJL auto-detection, which requires network access at runtime.
2. pre-install pytorch on the host and point DJL to pre-installed location
3. copy DJL's cached directory to your host manually

**Reply by user50** [reply]
thanks. works pretty good
```
FROM nvidia/cuda:11.7.0-cudnn8-devel-ubuntu20.04

ENV DEBIAN_FRONTEND=noninteractive
ENV TZ=UTC

RUN apt-get update && apt-get install -y \
    openjdk-11-jdk \
    curl \
    unzip \
    && rm -rf /var/lib/apt/lists/*

# Install PyTorch 1.13 and set DJL_PYTORCH_LIBRARY_PATH
ENV PYTORCH_VERSION=1.13.0
ENV PYTORCH_HOME=/opt/pytorch
ENV DJL_PYTORCH_LIBRARY_PATH=${PYTORCH_HOME}/lib
RUN mkdir -p ${PYTORCH_HOME} && \
    curl -sL https://download.pytorch.org/libtorch/cu117/libtorch-cxx11-abi-shared-with-deps-${PYTORCH_VERSION}%2Bcu117.zip -o libtorch.zip && \
    unzip libtorch.zip -d ${PYTORCH_HOME} && \
    rm libtorch.zip && \
    echo "export DJL_PYTORCH_LIBRARY_PATH=${PYTORCH_HOME}/lib" >> /etc/profile.d/pytorch.sh && \
    chmod +x /etc/profile.d/pytorch.sh
```
